admin/modules/tools/system_health.php

```diff
@@ -830,93 +830,42 @@
 		$table->output($lang->existing_db_backups);
 	}
 
-	if(is_writable(MYBB_ROOT.'inc/settings.php'))
-	{
-		$message_settings = "<span style=\"color: green;\">{$lang->writable}</span>";
-	}
-	else
-	{
-		$message_settings = "<strong><span style=\"color: #C00\">{$lang->not_writable}</span></strong><br />{$lang->please_chmod_777}";
-		++$errors;
-	}
-
-	if(is_writable(MYBB_ROOT.'inc/config.php'))
-	{
-		$message_config = "<span style=\"color: green;\">{$lang->writable}</span>";
-	}
-	else
-	{
-		$message_config = "<strong><span style=\"color: #C00\">{$lang->not_writable}</span></strong><br />{$lang->please_chmod_777}";
-		++$errors;
-	}
-
-	$uploadspath = $mybb->settings['uploadspath'];
-	if(is_writable($uploadspath))
-	{
-		$message_upload = "<span style=\"color: green;\">{$lang->writable}</span>";
-	}
-	else
-	{
-		$message_upload = "<strong><span style=\"color: #C00\">{$lang->not_writable}</span></strong><br />{$lang->please_chmod_777}";
-		++$errors;
-	}
-
-	$avataruploadpath = $mybb->settings['avataruploadpath'];
-	if(my_substr($avataruploadpath, 0, 1) == '.')
-	{
-		$avataruploadpath = MYBB_ROOT . $mybb->settings['avataruploadpath'];
-	}
-	if(is_writable($avataruploadpath))
-	{
-		$message_avatar = "<span style=\"color: green;\">{$lang->writable}</span>";
-	}
-	else
-	{
-		$message_avatar = "<strong><span style=\"color: #C00\">{$lang->not_writable}</span></strong><br />{$lang->please_chmod_777}";
-		++$errors;
-	}
+	$writables = array(
+		'config_file' => 'inc/config.php',
+		'settings_file' => 'inc/settings.php',
+		'file_upload_dir' => ltrim($mybb->settings['uploadspath'], MYBB_ROOT . '.'),
```

Comment:
This assumes that the current value of the setting (as entered by the admin) is a relative path (which has been converted in `inc/init.php` to an absolute one. If the current setting value, prior to conversion, is an absolute path though, then stripping MYBB_ROOT from the beginning is the wrong action to take here, and will display an incorrect value.

Reply:
No, if its absolute then only gets stripped otherwise there is no effect of the trim on the path string.

The array ensures to have all relative paths, because we are prepending `MYBB_ROOT` to all values altogether while checking writability just after:

`if (is_writable(MYBB_ROOT . $path))`

Reply:
What I mean is that let's say in `Server and Optimization Options` I set `Uploads Path` to the absolute path `/path/to/mybb/uploads`. When I open the `System Health` page, shouldn't I expect to see the value of `File Uploads Directory` set to the exact same value that I'd entered in those settings, namely `/path/to/mybb/uploads`? But with this PR, I don't. Instead, I see the relative path `./uploads`. Isn't the fact that admins weren't always seeing here the same value that they entered in settings the whole reason you created this PR, in which case, shouldn't this PR ensure that I see `/path/to/mybb/uploads` here in the scenario I've given?

Reply:
Again, those are not paths in the array. They are parts.
This action doesn't modify any path in system. It fetches the required part from a path. No matter what admin sets and what PR #4342 manipulates. It affects to all types of value provided.

If you are confused about the effect of the `ltrim` there, please see:

```php
echo ltrim('lairdshaw', 'airdl'); // shaw

echo ltrim('lairdshaw', 'dlaibarna'); // shaw
echo ltrim('barnardshaw', 'dlaibarna'); // shaw
```

Reply:
Sorry, may be I am not getting you right. Do you mean to say we should **display** all the absolute paths there?
The reason of my confusion is where exactly you have pointed. Thats not the visual part of the code, thats the working part. If you bother only about what admins see and what they should, you should have pointed here:

`$table->construct_cell('./' . trim($path, '/'));`

Reply:
I see I'd misunderstood how the second parameter of `ltrim()` worked, but now that I understand, my concerns are only magnified. e.g., if `$mybb->settings['uploadspath']` is set by the admin in the settings to `/path/to/some-other-dir/mybb/uploads`, then, if `MYBB_ROOT` is set to `/path/to/mybb`, then after the `ltrim` in this PR, the `uploadspath` setting will show up as the relative path ("part") as you want to call it, `some-other-dir/mybb/uploads`, which _looks_ like it is relative to the MyBB root but isn't.

In any case, you _do_ seem not to be getting me right. Please just answer this simple question:

If in `Server and Optimization Options` the admin sets `Uploads Path` to the absolute path, should the admin then see an absolute path for this setting on the `System Health` page? Why or why not?

Reply:
Now, as I partially understand what you say, still there is a regression.
You want if admins sets an absolute path in `Server and Optimization Options` he should see an absolute path there. Right?
But what if he sets a relative path instead? After your PR, it still displays an absolute path for `Uploads Path` and thats the regression.

Now if I change that particular line to `'file_upload_dir' => ltrim($mybb->settings['uploadspath'], '.'),` then I believe both of our views will meet. Of course I need to manipulate the checking part also.

Is that OK with you?

Reply:
> Now, as I partially understand what you say, still there is a regression.
> You want if admins sets an absolute path in `Server and Optimization Options` he should see an absolute path there. Right?

Yes. Especially if that absolute path is outside of the MyBB root. e.g., MyBB is installed at `/path/to/mybb`, and the uploads path is set to `/some/wild/and/crazy/path/to/uploads`.

> But what if he sets a relative path instead? After your PR, it still displays an absolute path for `Uploads Path` and thats the regression.

Right. I understand that that's what your PR is intended to fix. And I'm happy to work with you to see how we can do that.

> Now if I change that particular line to `'file_upload_dir' => ltrim($mybb->settings['uploadspath'], '.'),` then I believe both of our views will meet.

I don't think so, because the later line (currently #853):
```php
		$table->construct_cell('./' . trim($path, '/'));
```
still mistakenly ensures that an absolute path has its initial `/` converted to the apparently (but incorrectly) relative path `./`.

Reply:
> still mistakenly ensures that an absolute path has its initial / converted to the apparently (but incorrectly) relative path ./

As I said, **Of course I need to manipulate the checking part also.**

So, basically you mean to say I need to use the raw `$mybb->settings['uploadspath']` there. OK. so if I do that, my path:
`/home/eff.one/public_html/mtbbtest/uploads` (which I set just `./uploads` ) becomes
`/home/eff.one/public_html/mtbbtest/./uploads`

Still again as you said those two are equivalent and resolves to one same, there is much higher concern to worry.
1. I set my upload path as `./uploads`;
2. System converted it to `/home/eff.one/public_html/mtbbtest/./uploads`
3. I ran the writability check with raw `$mybb->settings['uploadspath']`. In some servers, this is the result:

![image](https://user-images.githubusercontent.com/21265591/115522108-73bef800-a2a9-11eb-9ef0-ab9f5fa916c2.png)

If I run the check with `/home/eff.one/public_html/mtbbtest/uploads` or `./uploads` it passes in both.

Now that I understand that the upload path may or may not be depended to and part of `MYBB_ROOT` the same applies to avatar upload path as well. That may or may not be depended to `upload parh` or even `MYBB_ROOT`. I need to check whether it has made depended in core in any way.

Reply:
To try to avoid further extensive back-and-forth, I suggest two possible resolutions to your concerns:

1. We don't absolutise the uploads path early on in processing, but instead wrap it in a "path-absolutising" function everywhere it is used.
2. We leave the early "absolutising" conversion of the uploads path as it is, but on the System Health page, we pull the original value out of the database for display and writeability checking.

Reply:
As for the avatar uploads path, as far as I understand, no changes have been made to it equivalent to those to the upload path in the recent PR #4342, but perhaps there should be.

Reply:
Alright, so I have rewritten the part, which I believe now OK and reflect correctly any presence of absolute path outside `MYBB_ROOT`. But yes, even if the admin sets an absolute path which is under `MYBB_ROOT` it will show up as relative, because it actually is.

Please let me know if the following is acceptable to you and I will push a commit:

```php

$writables = array(
	'config_file' => 'inc/config.php',
	'settings_file' => 'inc/settings.php',
	'file_upload_dir' => str_replace('/./', '/', $mybb->settings['uploadspath']),
	'avatar_upload_dir' => str_replace('/./', '/', $mybb->settings['avataruploadpath']),
	'language_files' => 'inc/languages',
	'backup_dir' => $config['admin_dir'] . '/backups',
	'cache_dir' => 'cache',
	'themes_dir' => 'cache/themes'
);

$errors = 0; // Reset errors
$table = new Table;
$table->construct_header($lang->checked_path);
$table->construct_header($lang->location);
$table->construct_header($lang->status_chmod, array('width' => 250));
foreach ($writables as $langvar => $path)
{
	$table->construct_cell("<strong>{$lang->{$langvar}}</strong>");
	$path = stripos($path, MYBB_ROOT) !== false ? $path : MYBB_ROOT . ltrim($path, './'); // TRIM IS REQUIRED BECAUSE AVATAR UPLOAD PATH IS STILL NOT ADDRESSED
	$table->construct_cell(str_replace(MYBB_ROOT, './', $path));
	if (is_writable($path))
	{
		$table->construct_cell("<span style=\"color: green;\">{$lang->writable}</span>");
	}
	else
	{
		$table->construct_cell("<strong><span style=\"color: #C00\">{$lang->not_writable}</span></strong><br />{$lang->please_chmod_777}");
		++$errors;
	}

	$table->construct_row();
}
```

And yes, avatar upload directory has to be addressed in the same way in your PR, I believe. Because its having the same issue.

Reply:
OK, after some brain storming I think we don't need to do the changes in avatar upload path, as the files are images and meant to be embedded in page: hence it has always to be relative / under `MYBB_ROOT`. Otherwise we need a proxy script in place to read the images, which is not present in current MyBB.

For avatars only what we can do is to explicitly declare in settings description that the path HAS TO BE relative.

Reply:
> Please let me know if the following is acceptable to you

I don't think it does the job it should do, no. To put my view in a nutshell: the `System Health` page should show beside `File Uploads Directory` _exactly_ what the admin entered for the `Uploads Path` setting in `Server and Optimization Options`. Your code does not achieve this, and makes some problematic transformations.

Reply:
I would like to know:

1. Which necessary part is removed from the path?
2. What is the procedure of reverse-manipulation of the base settings already manipulated by PR #4342 and overwritten in the settings value itself?
3. What are the test case scenarios that has failed with this PR?
4. Why we are at all considering some arbitrary absolute path out of domain root and in that case how we are managing the display of attachment image thumbnails in posts?

Reply:
> 1. Which necessary part is removed from the path?

It occurs in a scenario like this:

The admin sets `Uploads Path` in `Server and Optimization Settings` to:

`/some-dir/mybb/uploads`

when MYBB_ROOT is set to:

`/mybb`

In this scenario (though tested using different directories applicable on my dev machine), the new code you propose above for this PR will mistakenly display beside `File Uploads Directory` in `System Health`:

`/some-dir./uploads`

> 2. What is the procedure of reverse-manipulation of the base settings already manipulated by PR #4342 and overwritten in the settings value itself?

It is currently not possible to reliably reverse those changes, because it is impossible to know whether the setting was originally an absolute path or whether it was changed by PR #4342 from a relative path to an absolute one. Hence my suggestions above (I prefer the first): https://github.com/mybb/mybb/pull/4361#discussion_r617330863

> 3. What are the test case scenarios that has failed with this PR?

See number 1 above.

> 4. Why we are at all considering some arbitrary absolute path out of domain root

Because it's a valid value for the setting, and we should handle valid values for settings.

> and in that case how we are managing the display of attachment image thumbnails in posts?

I haven't confirmed this understanding, but my understanding is that neither full-sized nor thumbnail images are required to be web-accessible, only accessible to, and processed through, PHP via the web server, so that they can exist anywhere on the filesystem from/to which the web server can read and write, not necessarily under MYBB_ROOT. If I'm wrong about this, I hope you'll correct me.

Reply:
Attachment thumbnails are a very basic feature of MyBB and is there for ages. As per my understanding it was always considered that the upload attachments are under the domain root only.

If an admin deliberately inputs a settings like what you said, `/some-dir/mybb/uploads` then this is simply intentional and there are many settings like that in MyBB, infact anywhere - placing wrong settings simply fail the system. Thats the efficiency of Admin and thats why default values are there to hint. You can't ever make a setting input bullet-proof this way, no matter what you do. Speaking to the use case scenario you have pointed out that can be easily corrected in my above code and you also know how. In place of improving the effort your straight away rejection is stating that simply you are against this PR. 

I am leaving this PR here as a reference of the valuable discussion we already had for others.
You can close the issue / PR if you really feel its not a regression. Keeping your concept in consideration, if at all MyBB works this way then there is much more important core changes are required to support out-of-domain paths (starting with a proxy image dispatcher).

Sorry friend. I failed to conclude this on a common and agreed opinion.

Reply:
> As per my understanding it was always considered that the upload attachments are under the domain root only.

Well, in any case, deliberate or not, an uploads directory outside of the domain root works just fine (tested). This feature seems worth supporting given the extra security it provides given that the uploads directory can't then be accessed directly via the web.

> Speaking to the use case scenario you have pointed out that can be easily corrected in my above code and you also know how.

Correct. However, the original setting value (pre being altered by PR #4342) *can't* be determined, which is the real problem.

> you are against this PR.

I just don't understand the point of it. The only point to it that I can see would be to ensure that the _exact_ value of the `Uploads Path` setting entered by the admin in the settings is displayed in `System Health` given that that value is altered by PR #4342 - but it doesn't even achieve this.

So, what _is_ its point? To ensure that a relative path is displayed where possible - even if the admin entered an absolute path in the settings? If so, why?

> Sorry friend. I failed to conclude this on a common and agreed opinion.

That's OK. Disagreements happen, both to friends or otherwise.

Reply:
If #4370 is accepted, then we could go with this (tested):
```php
$writables = array(
	'config_file' => 'inc/config.php',
	'settings_file' => 'inc/settings.php',
	'file_upload_dir' => $mybb->settings['uploadspath'],
	'avatar_upload_dir' => $mybb->settings['avataruploadpath'],
	'language_files' => 'inc/languages',
	'backup_dir' => $config['admin_dir'] . '/backups',
	'cache_dir' => 'cache',
	'themes_dir' => 'cache/themes'
);

$errors = 0; // Reset errors
$table = new Table;
$table->construct_header($lang->checked_path);
$table->construct_header($lang->location);
$table->construct_header($lang->status_chmod, array('width' => 250));
foreach ($writables as $langvar => $path)
{
	$table->construct_cell("<strong>{$lang->{$langvar}}</strong>");
	$table->construct_cell($path);
	if (is_writable(mk_path_abs($path)))
	{
		$table->construct_cell("<span style=\"color: green;\">{$lang->writable}</span>");
	}
	else
	{
		$table->construct_cell("<strong><span style=\"color: #C00\">{$lang->not_writable}</span></strong><br />{$lang->please_chmod_777}");
		++$errors;
	}

	$table->construct_row();
}
```
Your thoughts?